### DISCUSSION TITLE
improvement streaming GCode

### DISCUSSION THREAD

**Discussion by liqijian101** [luc-github/ESP3D-TFT#75] [General]
There is a lag when printing the periphery, especially for circles
Is there any solution
https://github.com/user-attachments/assets/c70e3009-63f9-458c-a247-72a4bb87e5c5

**Comment by luc-github** [comment]
Yes improve streaming engine and do buffered streaming instead of simple ping/pong streaming.

It is in todo list but not on top list for the moment

**Comment by 3DSmitty** [comment]
Not sure if this helps. In CNC machining it's called Drip Feeding a program. One whole block (or line) of G code is sent to the controller's buffered memory at a time. The controller sends a signal to the external device (can be a PC connected over serial port or a networked server running DNC software) that it is ready for the next block of G code. The buffered memory size on the controller has to be large enough to handle the Look Ahead function (G103). Hopefully that helps.

**Reply by luc-github** [reply]
I am not aware of any GCode command `G103` in grbl/grblHAL

The buffered streaming is is based on what size in a available RX buffer reported in `?` report : https://github.com/gnea/grbl/blob/master/doc/markdown/settings.md#10---status-report-mask

The concept is to send as much command as the buffer can handle - while also counting number of commands sent and expected ack, for each ack received,  the streaming host check if new command fit the available size in rx buffer,  if yes it sent it one or more, if not streaming host wait for next ack to sent new command(s) 

This RX buffer need to take in account the `?` command and any others polling commands from the host, so it is a matter of counting chars and counting ack

**Reply by 3DSmitty** [reply]
I wasn't speaking about grbl but CNC machining in general (Haas, Fanuc) but that makes perfect sense how grbl uses the RX buffer.

**Comment by liqijian101** [comment]
MSK TFT has a good use effect, we should learn its way

**Reply by luc-github** [reply]
MKS do not full provide source code - only some libraries and striped source code, a lot is missing and it does not compile :
https://github.com/makerbase-mks/MKS-TFT70-Firmware/blob/master/Src/main.cpp#L980
https://github.com/makerbase-mks/MKS-TFT28-32-Firmware
https://github.com/makerbase-mks/MKS-TFT24-Firmware

BTT provide full code but I never had time to focus on the streaming code to write down the data flow and algorithm: https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/tree/master/TFT/src/User/API
